case-lib/lib.sh

```diff
@@ -693,6 +693,19 @@ ipc4_used()
     return 0
 }
 
+is_ipc4_zephyr(){
+    local firmware_path firmware_name znum
+    fw_path=/sys/module/snd_sof_pci/parameters/fw_path
+
+    firmware_path=$(cat $fw_path)
+    firmware_name=dsp_basefw.bin
+
+    ipc4_used || return 1
+
+    znum=$(strings "/lib/firmware/$firmware_path/$firmware_name" | grep -c -i zephyr)
+    test "$znum" -gt 10
```

Comment:
I recognize this code from the existing `is_zephyr()` function... which depends on the sof-logger .ldc file.

Please don't create a new, hidden and different `is_zephyr()` function here while keeping the old one. We need a single `is_zephyr()` function.

Either the old `is_zephyr()` function is still OK (because we still depend on sof-logger for DMA logs) and you call it here. Or it's not OK and you fix it; don't leave all the other `is_zephyr()` users broken.

Reply:
To find the firmware file you can try some combination of `./tools/sof-dump-status.py -P` and `./tools/sof-dump-status.py -p`. Again these are used in many other places, so if they don't work then them must be fixed.

@kv2019i , @plbossart, @ujfalusi  sof-test needs something like `/sys/module/snd_sof_pci/parameters/fw_path` but reliable that it can count on, can the kernel expose that? Same thing for the topology: up to now `sof-test/tools/sof-get-default-tplg.sh` has been grepping journalctl; that's a horrible and unreliable hack.


PS: actually the Linux kernel `request_firmware` API should expose all the firmware files it has  loaded from a filesystem, this sort of "manifest" seems like a bare minimum from a security perspective but that could take years.

Reply:
> Please don't create a new, hidden and different `is_zephyr()` function here while keeping the old one. We need a single `is_zephyr()` function.

Yes, I also considered to reuse `is_zephyr()` function, but it depends on the sof-logger .ldc file. But IPC4 firmware has 
 no ldc file.  if we want to reuse this function, we need to switch to using firmware binary instead of ldc file.

Reply:
> To find the firmware file you can try some combination of `./tools/sof-dump-status.py -P` and `./tools/sof-dump-status.py -p`

 `./tools/sof-dump-status.py -P` only works for IPC3 not IPC4. Part of path is fixed as `/lib/firmware/intel/sof/`,  it only check if the DUT is community board, if yes, the `fw_path` will be changed to `/lib/firmware/intel/sof/community`

Reply:
> can the kernel expose that? Same thing for the topology: up to now `sof-test/tools/sof-get-default-tplg.sh` has been grepping journalctl; that's a horrible and unreliable hack.

I totally agree with this proposal, then we can get the reliable fw_path from the journalctl.

Reply:
I agree that would be nice @keqiaozhang  and @marc-hb , but let's not block this PR while waiting for a new kernel interface. The hardcoded paths we have in kernel are quite sticky in the end. We can't just change them in the kernel anymore as there are countless system deployed with FW files in the place former kernel versions expected. So unless user has overriden with "fw_path", we can fairly reasonably expect what the firmware path will be.

Reply:
So here's how I think we can  get this merged soon with a minimum number of reasonable and manageable short-term hacks.

The most important thing is to break down `is_ipc4_zephyr()` into multiple, smaller functions. `is_ipc4_zephyr()` combines too many different logic pieces and decisions in a single function right now. Some pieces are short-term and others not. They must be separated into different functions.

- The easiest thing is to extract is a new function `is_fw_file_zephyr($filename) -> return true/false`

- Then another small function `cavs_filename($platform)` that prints the "semi-harcoded" `/lib/firmware/intel/avs/$platform/dsp_basefw.bin` If you want, let `fw_path` take precedence when defined but that's probably optional.

- Then combine the two functions above in a third new function `avs_is_zephyr()` that gets the platform from `sof-dump-status.py -p`


- Finally, in the logging code do something like:
```shell

	if ipc4_used && avs_is_zephyr; then

```

Reply:
I'm not sure I fully understand your point. 
IPC3 and IPC4 are using a completely different path:
IPC3: `/lib/firmware/intel/sof ` or `/lib/firmware/intel/sof/community/`
IPC4: `/lib/firmware/intel/sof-ipc4/$platform/community` or `/lib/firmware/intel/sof-ipc4/$platform/debug/`

So if we add a function like `is_fw_file_zephyr($filename) -> return true/false`, we first need to check `ipc4_used` or not. 
If we don't know the IPC type, then it's hard for us to know the specific `fw_path`, 
So the process should be:
1. check ipc type `ipc4_used`
2. get the  `fw_path`
3. check if the fw is zephyr

Almost the same steps as I did in function `is_ipc4_zephyr`.

Reply:
> So if we add a function like is_fw_file_zephyr($filename) -> return true/false, we first need to check ipc4_used or not.
> ...
> Almost the same steps as I did in function is_ipc4_zephyr

Yes but `is_ipc4_zephyr` is monolithic and unmaintainable. Break it down in smaller, modular, re-usable and maintainable functions.

Reply:
Sorry, I didn't figure out a better solution to break `is_ipc4_zephyr` down in smaller. 

I'm thinking to use firmware file to check zephyr in `is_zephyr`  instead of ldc file. 
We can refine this part after https://github.com/thesofproject/linux/issues/3867 is supported. 

Then combine these two functions `ipc4_used` +` is_zephyr`  to replace `is_ipc4_zephyr`.

any other suggestions?

Reply:
> I'm thinking to use firmware file to check zephyr in is_zephyr instead of ldc file.

Yes, this is what I suggested above: `is_fw_file_zephyr($filename)`. We unfortunately need two different `is_zephyr_XXX()` functions for now. Better make that obvious rather than hide it buried deep in a big, monolithic `is_ipc4_zephyr()` that looks like it's calling the usual `is_zephyr()` while doing something different behind the scenes.